## **1. Install the required libraries**

In [1]:
!pip install pymongo
!pip install pandas
!pip install numpy
!pip install scikit-learn
!pip install scikit-multilearn
!pip install statsmodels
!pip install matplotlib

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 648.9/648.9 kB 15.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 283.7/283.7 kB 31.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 89.4/89.4 kB 4.0 MB/s eta 0:00:00


## **2. Import the required libraries**

In [2]:
import pymongo
import json
import numpy as np
import pandas as pd
import datetime
from sklearn.preprocessing import MultiLabelBinarizer
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from skmultilearn.problem_transform import BinaryRelevance
from sklearn.preprocessing import MinMaxScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import matplotlib.pyplot as plt
from statsmodels.tsa.arima.model import ARIMA

## **3. Load Data fom MongoDB**

*   Accounts
*   Customers
*   Transactions


In [4]:
# Connect to MongoDB
client = pymongo.MongoClient('mongodb+srv://sakinahalizzah:Sakinah1234@clustersakinah.scfkjmg.mongodb.net/')
db = client["db_analytics"]
collection = db["accounts"]

# Retrieve data from collection
dataset = list(collection.find())

# Convert to DataFrame
df_accounts = pd.DataFrame(dataset)

# Print all rows of the DataFrame
print(df_accounts)

                           _id  account_id  limit  \
0     5ca4bbc7a2dd94ee5816238c      371138   9000   
1     5ca4bbc7a2dd94ee5816238d      557378  10000   
2     5ca4bbc7a2dd94ee5816238e      198100  10000   
3     5ca4bbc7a2dd94ee5816238f      674364  10000   
4     5ca4bbc7a2dd94ee58162390      278603  10000   
...                        ...         ...    ...   
1741  5ca4bbc7a2dd94ee58162a5c       89698  10000   
1742  5ca4bbc7a2dd94ee58162a5d      206062  10000   
1743  5ca4bbc7a2dd94ee58162a5e      684319  10000   
1744  5ca4bbc7a2dd94ee58162a5f      351063  10000   
1745  5ca4bbc7a2dd94ee58162a60      291224  10000   

                                               products  
0                        [Derivatives, InvestmentStock]  
1     [InvestmentStock, Commodity, Brokerage, Curren...  
2       [Derivatives, CurrencyService, InvestmentStock]  
3                                     [InvestmentStock]  
4                          [Commodity, InvestmentStock]  
...            

## Random Forest Classifier

In [58]:
# Convert product labels to binary representation
product_labels = pd.get_dummies(df_accounts['products'].apply(pd.Series).stack()).groupby(level=0).sum()

# Extract input features and labels
input_features = df_accounts[['account_id', 'limit']]
labels = product_labels.values

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(input_features, labels, test_size=0.2, random_state=42)

# Normalize the input features using min-max scaling
scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Train the Random Forest Classifier
rf_classifier = RandomForestClassifier()
rf_classifier.fit(X_train_scaled, y_train)

# Predict labels for the testing data
predictions = rf_classifier.predict(X_test_scaled)


# Print the predicted labels
predicted_labels = pd.DataFrame(predictions, columns=product_labels.columns)
print(predicted_labels)

     Brokerage  Commodity  CurrencyService  Derivatives  InvestmentFund  \
0            0          0                0            1               1   
1            0          1                1            0               1   
2            1          0                1            1               0   
3            1          0                0            1               0   
4            1          1                1            0               0   
..         ...        ...              ...          ...             ...   
345          0          0                0            1               1   
346          1          0                0            0               0   
347          1          0                1            0               1   
348          0          1                0            0               1   
349          0          0                1            0               1   

     InvestmentStock  
0                  1  
1                  1  
2                  1  
3      

In [65]:
# Calculate the sum of each product
sum_by_product = predicted_labels.sum(axis=0)

# Sort the sums in descending order
sum_by_product_sorted = sum_by_product.sort_values(ascending=False)

# Print the sorted sums
print(sum_by_product_sorted)


InvestmentStock    350
Brokerage          166
CurrencyService    161
Derivatives        151
InvestmentFund     148
Commodity          142
dtype: int64


## Linear Regression

### to predict the number of accounts a customer has based on their age.

In [5]:
collection = db["customers"]

# Retrieve data from collection
dataset = list(collection.find())

# Convert to DataFrame
df_customers = pd.DataFrame(dataset)

# Print all rows of the DataFrame
print(df_customers)

                          _id           username                  name  \
0    5ca4bbcea2dd94ee58162a68            fmiller         Elizabeth Ray   
1    5ca4bbcea2dd94ee58162a69   valenciajennifer         Lindsay Cowan   
2    5ca4bbcea2dd94ee58162a6a         hillrachel       Katherine David   
3    5ca4bbcea2dd94ee58162a6b       serranobrian       Leslie Martinez   
4    5ca4bbcea2dd94ee58162a6c      charleshudson         Brad Cardenas   
..                        ...                ...                   ...   
495  5ca4bbcea2dd94ee58162c5a     amandawilliams          Brandy Huang   
496  5ca4bbcea2dd94ee58162c5b  stricklandjeffery          Xavier Myers   
497  5ca4bbcea2dd94ee58162c5c          smcintyre  Christopher Lawrence   
498  5ca4bbcea2dd94ee58162c5d            qknight        Gabriel Romero   
499  5ca4bbcea2dd94ee58162c5e             ecasey     Brandon Contreras   

                                               address           birthdate  \
0            9286 Bethany Glens\n

In [12]:
# Convert birthdate to age
current_year = datetime.datetime.now().year
df_customers['birthdate'] = pd.to_datetime(df_customers['birthdate'], unit='ms')  # Convert the birthdate to datetime
df_customers['age'] = current_year - df_customers['birthdate'].dt.year

# Count total accounts for each customer
df_customers['total_accounts'] = df_customers['accounts'].apply(lambda x: len(x))

# Display the selected columns in a table format
print(df_customers[['name', 'birthdate', 'age', 'total_accounts']].to_string(index=False))

                   name           birthdate  age  total_accounts
          Elizabeth Ray 1977-03-02 02:20:31   46               6
          Lindsay Cowan 1994-02-19 23:46:27   29               1
        Katherine David 1988-06-20 22:15:34   35               5
        Leslie Martinez 1974-11-26 14:30:20   49               2
          Brad Cardenas 1977-05-06 21:57:35   46               5
           Natalie Ford 1996-09-13 17:14:27   27               2
            Dana Clarke 1969-06-21 02:39:20   54               3
           Gary Nichols 1993-02-25 17:37:29   30               6
             John Parks 1993-03-13 11:37:34   30               2
      Jennifer Lawrence 1972-11-10 11:01:08   51               3
          James Sanchez 1973-01-13 16:17:26   50               1
           Ashley Lopez 1989-11-24 16:12:54   34               2
       Dr. Angela Brown 1977-06-19 20:35:52   46               1
              John Vega 1983-09-21 03:22:18   40               4
           Lauren Clark 1

In [18]:
# Find the customer with the highest and lowest total accounts based on age
highest_accounts_customer = df_customers.loc[df_customers['total_accounts'].idxmax()]
lowest_accounts_customer = df_customers.loc[df_customers['total_accounts'].idxmin()]

# Calculate the total number of customers with the highest and lowest total accounts
total_highest_accounts = len(df_customers[df_customers['total_accounts'] == highest_accounts_customer['total_accounts']])
total_lowest_accounts = len(df_customers[df_customers['total_accounts'] == lowest_accounts_customer['total_accounts']])

# Calculate the average age of customers with the highest and lowest total accounts
average_age_highest_accounts = df_customers[df_customers['total_accounts'] == highest_accounts_customer['total_accounts']]['age'].mean()
average_age_lowest_accounts = df_customers[df_customers['total_accounts'] == lowest_accounts_customer['total_accounts']]['age'].mean()

# Print the total number of customers and average age
print("Total number of customers with the highest total accounts:", total_highest_accounts)
print("Total number of customers with the lowest total accounts:", total_lowest_accounts)
print("Average age of customers with the highest total accounts:", average_age_highest_accounts)
print("Average age of customers with the lowest total accounts:", average_age_lowest_accounts)


Total number of customers with the highest total accounts: 83
Total number of customers with the lowest total accounts: 83
Average age of customers with the highest total accounts: 39.0
Average age of customers with the lowest total accounts: 41.21686746987952


In [21]:
# Convert birthdate to age
current_year = datetime.datetime.now().year
df_customers['age'] = current_year - pd.to_datetime(df_customers['birthdate']).dt.year

# Prepare features and target
features = df_customers['age'].values.reshape(-1, 1)
target = df_customers['accounts'].apply(len).values

# Split data into train and test sets
X_train, X_test, y_train, y_test = train_test_split(features, target, test_size=0.2, random_state=42)

# Train the linear regression model
regressor = LinearRegression()
regressor.fit(X_train, y_train)

# Make predictions on the test set
predictions = regressor.predict(X_test)

# Calculate mean squared error
mse = mean_squared_error(y_test, predictions)
print("Mean Squared Error:", mse)

Mean Squared Error: 2.4954232672947128


the Mean Squared Error value of 2.4954 indicates that, on average, the squared difference between the predicted number of accounts and the actual number of accounts is approximately 2.4954. A lower Mean Squared Error value indicates better model performance, as it represents a smaller average squared difference between the predicted and actual values.